tests/integration/test_recipe.py

```diff
@@ -1841,6 +1841,184 @@ def test_extract_shape_raises(tmp_path, patched_datafinder, config_user,
     assert invalid_arg in exc.value.failed_tasks[0].message
 
 
+def _test_output_product_consistency(products, preprocessor, statistics):
+    product_out = defaultdict(list)
+
+    for i, product in enumerate(products):
+        settings = product.settings[preprocessor]
+        output_products = settings['output_products']
+
+        for identifier, statistic_out in output_products.items():
+            for statistic, preproc_file in statistic_out.items():
+                product_out[identifier, statistic].append(preproc_file)
+
+    # Make sure that output products are consistent
+    for (identifier, statistic), value in product_out.items():
+        assert statistic in statistics
+        assert len(set(value)) == 1, 'Output products are not equal'
+
+    return product_out
+
+
+def test_ensemble_statistics(tmp_path, patched_datafinder, config_user):
+    statistics = ['mean', 'max']
+    diagnostic = 'diagnostic_name'
+    variable = 'pr'
+    preprocessor = 'ensemble_statistics'
+
+    content = dedent(f"""
+         preprocessors:
+           default: &default
+             custom_order: true
+             area_statistics:
+               operator: mean
+             {preprocessor}:
+               statistics: {statistics}
+
+         diagnostics:
+           {diagnostic}:
+             variables:
+               {variable}:
+                 project: CMIP5
+                 mip: Amon
+                 start_year: 2000
+                 end_year: 2002
+                 preprocessor: default
+                 additional_datasets:
+                   - {{dataset: CanESM2, exp: [historical, rcp45],
+                     ensemble: "r(1:2)i1p1"}}
+                   - {{dataset: CCSM4, exp: [historical, rcp45],
+                     ensemble: "r(1:2)i1p1"}}
+             scripts: null
+    """)
+
+    recipe = get_recipe(tmp_path, content, config_user)
+    variable = recipe.diagnostics[diagnostic]['preprocessor_output'][variable]
+    datasets = set([var['dataset'] for var in variable])
+    task = next(iter(recipe.tasks))
+
+    products = task.products
+    product_out = _test_output_product_consistency(products, preprocessor,
+                                                   statistics)
+
+    assert len(product_out) == len(datasets) * len(statistics)
+
+    task._initialize_product_provenance()
+    assert next(iter(products)).provenance is not None
+
+
+def test_multi_model_statistics(tmp_path, patched_datafinder, config_user):
+    statistics = ['mean', 'max']
+    diagnostic = 'diagnostic_name'
+    variable = 'pr'
+    preprocessor = 'multi_model_statistics'
+
```

Comment:
you can use the list of all sorts of stuff decorator `@pytest.mark.parametrize()` to throw in all sorts of parameters for the test, no need to list them internally and go one by one

Reply:
I am not sure this is written in this way, I did not write this test sorry.

Reply:
I don't think it's going one by one. I guess it was written like this because those variables are reused multiple times (in the recipe definition and in the check). Indeed, since it's written like this, you can now easily extend the tests with multiple combinations of variables etc., but since it's only one test now, adding `parametrize` is unnecessary.